chore: automate project standards

kriscoleman opened the pull request:
## This PR

- Implements Lefthook for easy git hooks, w/ configuration to:
  - Ensures`go fmt` as a pre-commit hook with `stage_fixed: true` (it will automatically stage the fixes), preventing pushes with bad code formatting. 
  - Ensures `make generate-docs` is executed pre-push, preventing a push with forgotten document generation updates
  - Ensures `go test` is executed and tests pass pre-push, preventing a push with broken tests
- Updates the `README.md` to reflect the latest Lefthook setup and usage instructions.

### Notes

- I find these kinds of automation to help tremendously - when switching projects and adhering to different standards, having this automation in place can make jumping around seamless, and prevents wasting time when you forget to handle checks before a push. 
- The Lefthook configuration now ensures consistent formatting and documentation generation across the team.
- Developers can manually run hooks using `lefthook run pre-commit` or `lefthook run pre-push`.

### Follow-up Tasks

- Monitor team feedback on the updated Lefthook setup.
- Consider adding additional hooks if needed, such as security checks, or `go vet`.

### How to test

1. Install Lefthook using `brew install lefthook`.
2. Run `lefthook install` to opt-in and set up the hooks.
3. Test the pre-commit hook:
   - Make a change to a Go file and commit it.
   - Verify that `go fmt` runs and stages any changes automatically.
4. Test the pre-push hook:
   - Introduce a change to the auto-generated documentation, like delete an entire section. Commit the change.
   - Attempt to push and verify that the push is blocked.
   - Break a test on purpose and attempt to push and verify that the push is blocked if tests fail.

<img width="1178" alt="image" src="https://github.com/user-attachments/assets/388e3350-47f7-447a-83c2-0b648393d4b3" />

beeme1mr commented on README.md:
````diff
@@ -220,6 +220,53 @@ flowchart LR
   config --> args
 ```
 
+## Setting Up Lefthook
````
Please move this to the contributing docs.

kriscoleman replied:
✅ done

beeme1mr commented on lefthook.yml:
It may be worth adding a comment at the top of the file that directs the user towards the relevant section in the contributing docs.

kriscoleman replied:
✅ done

beeme1mr reviewed (approved):
This is great, thanks!

kriscoleman commented:
> Hey @kriscoleman, please ping me when the merge conflicts and readme changes have been addressed so I can merge the PR. Thanks!

Sounds great @beeme1mr 
I'll wait until after #97 is merged, then rebase this.

kriscoleman commented:
@beeme1mr rebased this and addressed your feedback on docs. I also dropped the fmt commit, because that got merged in the other PR. I reran all hooks and they passed locally and no changes were generated.
